toc/rfc/rfc-draft-pcap-bosh.md

```diff
@@ -2,60 +2,224 @@
+- BOSH instance group(s), and optionally instance IDs
+- [pcap-filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) to apply to the captured traffic.
 
-PCAP CLI will form a capture request and send it to PCAP API. The capture will then take place in real-time similar to running [tcpdump](https://www.tcpdump.org/manpages/tcpdump.1.html) in a distributed fashion across multiple machines. The PCAP API will contact PCAP AGENT on all nodes involved and trigger a network capture. The PCAP AGENTs will concurrently stream all captured traffic back to PCAP AGENT where all streams are merged into one contiguous stream that is sent back to the PCAP CLI where it is then stored as a PCAP FILE on disk.
+Capture targets and their IP addresses are then identified automatically based on looking up metadata from the BOSH Director using these parameters.
 
-All communication between PCAP CLI, PCAP API and PCAP AGENT(s) is carried out via gRPC over HTTP/2.
-
-Trust is established as follows:
-1. PCAP CLI uses the BOSH CLI `access_token` and sends them to PCAP API
-2. PCAP API verifies `access_token` against BOSH UAA, verifies the token scope and establishes mTLS with PCAP AGENT
-3. PCAP AGENT only accepts mTLS connections from PCAP AGENT via a trusted CA
+A single, combined stream of capture data is then provided as output and can be written to file or piped into another command.
 
 ## Problem
 
-The current architecture for "PCAP for BOSH" is depicted as follows:
-
-![PCAP BOSH Architecture](rfc-draft-pcap-bosh/tcpdump-for-bosh.svg)
+Distributed, and particularly networked systems are hard to debug. Software developers and operators are not necessarily the same people, in some cases not even the same companies.
 
+Network traffic can be reasonably easily captured via tools such as `tcpdump` on local machines. In order to get a unified view of the network traffic of relevant nodes, those captures need to be retrieved, merged and analysed.
 
-There are several flaws with this architecture that are the result of missing integration:
-1. The PCAP BOSH CLI does not share context with BOSH CLI (because unlike CF CLI, BOSH CLI has no plugin mechanism).
-2. PCAP BOSH CLI reads the user's `access_token` and `refresh_token`, talks to BOSH UAA to get a new `access_token` using the `refresh_token` and writes both back to `.bosh/config` - a file which is also accessed by BOSH CLI.
-3. The `access_token` is then sent to PCAP API which uses it as a credential to talk to BOSH DIRECTOR
-4. PCAP API needs to speak to BOSH DIRECTOR in order to find the endpoints (IP:PORT) of each PCAP AGENT needed for the capture.
-5. In order to make this feature work, PCAP AGENT needs to be available on all BOSH managed VMs. The only way today would be to use it as an [addon](https://bosh.io/docs/runtime-config/#addons)
+BOSH does not have a user or operator friendly way for resolving targets, or capturing, filtering and merging network traffic.
```

Comment:
This always throws me off, to me this list sounds like the BOSH can't provide any of the listed operations on their own, but resolving targets is just `bosh vms` isn't it? It's the combination of "resolving targets, and capturing, filtering and merging traffic", isn't it (although the double and doesn't sound great)?

Reply:
good catch. This sentence is confusing. I will refine it. As you say, BOSH is particularly good at identifying targets, etc. but tcpdump isn't.

The same mechanism as for BOSH instance resolution should be used, i.e. when not providing an instance group on a deployment it would take all instances, e.g.

```shell```
# capture all gorouter instances
bosh pcap -d cf gorouter

# capture all HAProxy instances on instance group `ha_proxy_z1`
bosh pcap -d haproxy ha_prox_z1

# capture the first NATS server
bosh pcap -d cf nats/0
```

Reply:
LGTM!